src/runtime_src/core/pcie/driver/linux/xocl/subdev/ert_user.c

```diff
@@ -0,0 +1,986 @@
+			continue;
+		xcmd->cb.notify_host(xcmd, status);
+		list_del(&ecmd->list);
+		xcmd->cb.free(xcmd);
+		ert_user_free_cmd(ecmd);
+		--(*len);
+	}
+}
+
+/* Use for flush submit queue */
+static void
+flush_submit_queue(struct xocl_ert_user *ert_user, u32 *len, int status, void *client)
+{
+	struct kds_command *xcmd;
+	struct ert_user_command *ecmd;
+	u32 i = 0;
+	unsigned long flags;
+
+
+	spin_lock_irqsave(&ert_user->sq_lock, flags);
+	for ( i = 0; i < ERT_MAX_SLOTS; ++i ) {
+		ecmd = ert_user->submit_queue[i];
+		if (ecmd) {
+			xcmd = ecmd->xcmd;
+			if (client && client != xcmd->client)
+				continue;
+			xcmd->cb.notify_host(xcmd, status);
+			xcmd->cb.free(xcmd);
+			ert_user->submit_queue[i] = NULL;
+			ert_user_free_cmd(ecmd);
+			--ert_user->num_sq;
+		}
+	}
+	spin_unlock_irqrestore(&ert_user->sq_lock, flags);
+}
+/*
+ * release_slot_idx() - Release specified slot idx
+ */
+static void
+ert_release_slot_idx(struct xocl_ert_user *ert_user, unsigned int slot_idx)
+{
+	clear_bit(slot_idx, ert_user->slot_status);
+}
+
+/**
+ * release_slot() - Release a slot index for a command
+ *
+ * Special case for control commands that execute in slot 0.  This
+ * slot cannot be marked free ever.
+ */
+static void
+ert_release_slot(struct xocl_ert_user *ert_user, struct ert_user_command *ecmd)
+{
+	if (ecmd->slot_idx == no_index)
+		return;
+
+	if (cmd_opcode(ecmd) == OP_CONFIG) {
+		ERTUSER_DBG(ert_user, "do nothing %s\n", __func__);
+		ert_user->ctrl_busy = false;
+		ert_user->config = true;	
+	} else {
+		ert_release_slot_idx(ert_user, ecmd->slot_idx);
+	}
+	ecmd->slot_idx = no_index;
+}
+
+/**
+ * process_ert_cq() - Process completed queue
+ * @ert_user: Target XRT CU
+ */
+static inline void process_ert_cq(struct xocl_ert_user *ert_user)
+{
+	struct kds_command *xcmd;
+	struct ert_user_command *ecmd;
+
+	if (!ert_user->num_cq)
+		return;
+
+	ERTUSER_DBG(ert_user, "-> %s\n", __func__);
+	/* Notify host and free command */
+	ecmd = list_first_entry(&ert_user->cq, struct ert_user_command, list);
+	xcmd = ecmd->xcmd;
+	ERTUSER_DBG(ert_user, "%s -> ecmd->slot_idx %d xcmd%p\n", __func__, ecmd->slot_idx, xcmd);
+	ert_release_slot(ert_user, ecmd);
+	list_del(&ecmd->list);
+	xcmd = ecmd->xcmd;
+	xcmd->cb.notify_host(xcmd, KDS_COMPLETED);
+	xcmd->cb.free(xcmd);
+	ert_user_free_cmd(ecmd);
+	--ert_user->num_cq;
+	ERTUSER_DBG(ert_user, "<- %s\n", __func__);
+}
+
+/**
+ * mask_idx32() - Slot mask idx index for a given slot_idx
+ *
+ * @slot_idx: Global [0..127] index of a CQ slot
+ * Return: Index of the slot mask containing the slot_idx
+ */
+static inline unsigned int
+mask_idx32(unsigned int idx)
+{
+	return idx >> 5;
+}
+
+static irqreturn_t
+ert_user_versal_isr(int irq, void *arg)
+{
+	struct xocl_ert_user *ert_user = (struct xocl_ert_user *)arg;
+	xdev_handle_t xdev = xocl_get_xdev(ert_user->pdev);
+
+	ERTUSER_DBG(ert_user, "-> %s %d\n", __func__, irq);
+
+	if (ert_user) {
+		xocl_mailbox_versal_handle_intr(xdev);
+#if 0
```

Comment:
@larry9523 , how does versal handle the isr?
do we have to read the csr register from host side?

Reply:
No, on versal, no csr read. 
1. there is only one interrupt to mb_scheduler
2. interrupt handler will wake up scheduler
3. scheduler will read mailbox between device and host to figure out what cq slot is done.